<HTML>
<HEAD>
<META HTTP-EQUIV="CONTENT-TYPE" CONTENT="TEXT/HTML; CHARSET=UTF-8">
<TITLE>_matherr - Process Math Library Errors</TITLE>
</HEAD>
<BODY>
<H1>_matherr - Process Math Library Errors</H1><!-- entering slot 195 -->_matherr - Process Math Library Errors<!-- lm: 0x2 1 -->
<UL>
<P>
Syntax
<P>
<!-- lm: 0x2 10 -->
<UL>
<PRE>#include &lt;math.h&gt;
int _matherr(struct exception *<I>x</I>);
</PRE>
<!-- lm: 0x2 1 -->
</UL>Description
<P>
<!-- lm: 0x2 10 -->
<UL>
<P>
<LI>matherr processes errors generated by the functions in the math library.
The math functions call _matherr whenever they detect an error. The _matherr
function supplied with<I> The Developer's Toolkit</I> library performs no
error handling and returns 0 to the calling function.
<P>
You can provide a different definition of _matherr to carry out special
error handling. Be sure to use the /NOE link option, if you are using your
own _matherr function. For<I> The Developer's Toolkit</I> compiler, you
can use the /B option on the icc command line to pass the /NOE option to
the linker, as in the following:
<PRE>   icc /B"/NOE" <I>yourfile.c</I>
</PRE>

<P>
When an error occurs in a math routine, _matherr is called with a pointer
to the following structure (defined in &lt;math.h&gt;) as an argument:
<PRE>   struct exception {
       int type;
       char *name;
       double arg1, arg2, retval;
   };
</PRE>

<P>
The type field specifies the type of math error. It has one of the following
values, defined in &lt;math.h&gt;:  compact break=fit.<B><!-- lm: 0x2 10 -->
<P>
Value<!-- lm: 0x2 25 -->
<UL> Meaning</B><!-- lm: 0x2 10 -->
</UL> DOMAIN
<!-- lm: 0x2 25 -->
<UL>Argument domain error<!-- lm: 0x2 10 -->
</UL> OVERFLOW
<!-- lm: 0x2 25 -->
<UL>Overflow range error<!-- lm: 0x2 10 -->
</UL> UNDERFLOW
<!-- lm: 0x2 25 -->
<UL>Underflow range error<!-- lm: 0x2 10 -->
</UL> TLOSS
<!-- lm: 0x2 25 -->
<UL>Total loss of significance<!-- lm: 0x2 10 -->
</UL> PLOSS
<!-- lm: 0x2 25 -->
<UL>Partial loss of significance<!-- lm: 0x2 10 -->
</UL> SING
<!-- lm: 0x2 25 -->
<UL>Argument singularity.<!-- lm: 0x2 10 -->
</UL>
<P>
PLOSS is provided for compatibility with other compilers;<I> The Developer's
Toolkit</I> does not generate this value.
<P>
The name is a pointer to a null-ended string containing the name of the
function that caused the error. The arg1 and arg2 fields specify the argument
values that caused the error. If only one argument is given, it is stored
in arg1. The retval is the default return value; you can change it.<!-- lm: 0x2 1 -->
</UL> Returns

<P>
<!-- lm: 0x2 10 -->
<UL>
<P>
The return value from _matherr must specify whether or not an error actually
occurred. If _matherr returns 0, an error message appears, and errno is
set to an appropriate error value. If _matherr returns a nonzero value,
no error message appears and errno remains unchanged.<!-- lm: 0x2 1 -->
</UL> Example
Code
<P>
<!-- lm: 0x2 10 -->
<UL>
<P>
This example provides a _matherr function to handle errors from the log
or log10 functions. The arguments to these logarithmic functions must be
positive double values. _matherr processes a negative value in an argument
(a domain error) by returning the log of its absolute value. It suppresses
the error message normally displayed when this error occurs.  If the error
is a zero argument or if some other routine produced the error, the example
takes the default actions.
<PRE>#include &lt;math.h&gt;#include &lt;string.h&gt;
#include &lt;stdio.h&gt;
#include &lt;stdlib.h&gt;

int main(void)
{
   int value;

   printf("Trying to evaluate log10(-1000.00) will create a math exception.\n");
   value = log10(-1000.00);
   printf("The _matherr() exception handler evaluates the expression to\n");
   printf("log10(-1000.00) = %d\n", value);
   return 0;

   /****************************************************************************
      The output should be:

      Trying to evaluate log10(-1000.00) will create a math exception.
      inside _matherr
      The _matherr() exception handler evaluates the expression to
      log10(-1000.00) = 3
   ****************************************************************************/
}

int _matherr(struct exception *x)
{
   printf("inside _matherr\n");
   if (DOMAIN == x-&gt;type) {
      if (0 == strcmp(x-&gt;name, "log")) {
         x-&gt;retval = log(-(x-&gt;arg1));
         return EXIT_FAILURE;
      }
      else
         if (0 == strcmp(x-&gt;name, "log10")) {
            x-&gt;retval = log10(-(x-&gt;arg1));
            return EXIT_FAILURE;
         }
   }
   return 0;                                  /* Use default actions          */
}
</PRE>
<!-- lm: 0x2 1 -->
</UL>Related Information
<P>
<!-- lm: 0x2 10 -->
<UL><!-- lm: 0x2 12 -->
<UL>
<P>
<LI><!-- lm: 0x2 15 --><A HREF="179_L2_logCalculateNaturalL.html">log</A><!-- lm: 0x2 12 -->
<BR>
<LI><!-- lm: 0x2 15 --><A HREF="180_L2_log10CalculateBase10.html">log10
</A><!-- lm: 0x2 10 -->
</UL>

<P><HR>

<A HREF="185_L2_mallocReserveStorage.html">[Back: malloc - Reserve Storage Block]</A> <BR>
<A HREF="187_L2_maxReturnLargerofTwo.html">[Next: max - Return Larger of Two Values]</A> 
</BODY>
</HTML>
